### DISCUSSION TITLE
Synchronize attempt results in "External security challenge not satisfied."

### DISCUSSION THREAD

**Discussion by gitkevin** [abraunegg/onedrive#2420] [Q&A] [answered]
I'm using onedrive version 2.4.13, and that's worked very well for at least 6 months by now. I've had to resynchronize my login credentials a few times now. That's always worked for me, until now (June 15, 2023). My synchronization process went something like below.

```
onedrive --synchronize --verbose
Using 'user' Config Dir: /home/myaccountname/.config/onedrive
Using 'system' Config Dir: /etc/onedrive
Configuration file successfully loaded
Initializing the OneDrive API ...
Configuring Global Azure AD Endpoints
Authorize this app visiting:

https://login.microsoftonline.com/common/oauth2/v2.0/authorize?client_id=NOT_MY_REAL_INFO=Files.ReadWrite%20Files.ReadWrite.all%20Sites.Read.All%20Sites.ReadWrite.All%20offline_access&response_type=code&redirect_uri=https://login.microsoftonline.com/common/oauth2/nativeclient

Enter the response uri: https://login.microsoftonline.com/common/oauth2/nativeclient?code=not_my_real_code&session_state=ac960913-e047-47f2-a330-ac25db283535

ERROR: Microsoft OneDrive API returned an error with the following message:
  Error Message:    HTTP request returned status code 400 (Bad Request)
  Error Reason:     AADSTS50158: External security challenge not satisfied. User will be redirected to another page or authentication provider to satisfy additional authentication challenges.
Trace ID: a5261756-c6a8-41d4-af7a-2e81caf11900
Correlation ID: 4975f002-2404-4688-9007-c0e03b5ffb32
Timestamp: 2023-06-15 16:07:59Z
  Calling Function: acquireToken()

Opening the item database ...
All operations will be performed in: /home/myaccountname/OneDrive

ERROR: Microsoft OneDrive API returned an error with the following message:
  Error Message:    HTTP request returned status code 400 (Bad Request)
  Error Reason:     AADSTS900144: The request body must contain the following parameter: 'refresh_token'.
Trace ID: 8c8ac355-c782-4bed-978e-811324e81700
Correlation ID: 75da0425-7368-452f-a10e-522c87add5c7
Timestamp: 2023-06-15 16:08:00Z
  Calling Function: acquireToken()


ERROR: Microsoft OneDrive API returned an error with the following message:
  Error Message:    HTTP request returned status code 401 (Unauthorized)
  Error Reason:     Access token is empty.
  Error Timestamp:  2023-06-15T16:08:00
  API Request ID:   4f4b37aa-4832-45ed-afbf-c64aa0b96aaa
  Calling Function: init()


ERROR: Check your configuration as your refresh_token may be empty or invalid. You may need to issue a --logout and re-authorise this client.

Segmentation fault (core dumped)
```

The seg fault makes me think that a shared object library used by onedrive got updated in an incompatible way. 

My configuration, as told to me by "onedrive --synchronize --verbose" is:

```
Configuration file successfully loaded
onedrive version                       = v2.4.13-1build1
Config path                            = /home/myaccountname/.config/onedrive
Config file found in config path       = true
Config option 'check_nosync'           = false
Config option 'sync_dir'               = /home/myaccountname/OneDrive
Config option 'skip_dir'               = 
Config option 'skip_file'              = ~*|.~*|*.tmp
Config option 'skip_dotfiles'          = false
Config option 'skip_symlinks'          = false
Config option 'monitor_interval'       = 300
Config option 'min_notify_changes'     = 5
Config option 'log_dir'                = /var/log/onedrive/
Config option 'classify_as_big_delete' = 1000
Config option 'upload_only'            = false
Config option 'no_remote_delete'       = false
Config option 'remove_source_files'    = false
Config option 'sync_root_files'        = false
Selective sync 'sync_list' configured  = true
sync_list contents:
# sync_list supports comments
#
# The ordering of entries is highly recommended - exclusions before inclusions
#
# Exclude unshared content
!/MyNotes/
#
# Include myproj
!/myproj/Development/
/myproj
/keepout
Business Shared Folders configured     = false
```

There's another possibility for why this might not be working. In the past I've always done all the steps on one computer. Today I'm working remotely. I reach the Linux machine running onedrive via ssh and run onedrive in a terminal shell. The URLs that get copied and pasted are used in a browser local to me. That means that to the onedrive server the onedrive client will be coming from a specific IP address, and the URL exchange via the browser will be coming from a totally different IP address. If I were coding onedrive, I think I'd disallow it as it might look like an attack. Any comments on this? When I'm back in a location where only one IP address is involved, then I'll try this again and report back.

I'm running LinuxMint 21.1. 

Is my only, or best alternative here to grab the tar.gz source code and attempt to build and install that?   If there's another alternative to making this work again, I'm open to all suggestions. Of course, I'll report back when I get something working.

Thanks all.

**Comment by abraunegg** [comment, marked as answer]
@gitkevin 
Firstly you are running an unsupported application version. You are running an application version from Jul 14, 2021 and there have been significant bugs that have been fixed since then

Please read: 
* https://github.com/abraunegg/onedrive#supported-application-version
* https://github.com/abraunegg/onedrive/releases
* https://github.com/abraunegg/onedrive/blob/master/docs/INSTALL.md
* https://github.com/abraunegg/onedrive/blob/master/docs/ubuntu-package-install.md

> `AADSTS50158: External security challenge not satisfied. User will be redirected to another page or authentication provider to satisfy additional authentication challenges.`
> ...
> That means that to the onedrive server the onedrive client will be coming from a specific IP address, and the URL exchange via the browser will be coming from a totally different IP address. If I were coding onedrive, I think I'd disallow it as it might look like an attack. Any comments on this? When I'm back in a location where only one IP address is involved, then I'll try this again and report back.

Your organisation is restricting where you can connect to OneDrive from.

If you have a work VPN when working remote, you need to update your VPN connection to tunnel all your OneDrive access via that - so you may need to speak to your IT department.

**Reply by gitkevin** [reply]
@abraunegg
I uninstalled the obsolete onedrive and rebuilt onedrive per instructions for building from source. That was remarkably simple, and I would encourage people to not fear those steps. The resulting onedrive binary functioned well enough to let me look at the configuration.

The server I'm attending to is command-line only, meaning I don't get to use a plush GUI browser. I'm on the trail of a text browser with JavaScript support. Once I figure that out, then I'll try to synchronize onedrive.

**Reply by abraunegg** [reply]
On Linux Mint you should be using the OpenSuSE packages ... you should not have to build the client from source for Linux Mint.

**Comment by gitkevin** [comment]
I said I'd write back about the following.

> There's another possibility for why this might not be working. In the past I've always done all the steps on one computer. Today I'm working remotely. I reach the Linux machine running onedrive via ssh and run onedrive in a terminal shell. The URLs that get copied and pasted are used in a browser local to me. That means that to the onedrive server the onedrive client will be coming from a specific IP address, and the URL exchange via the browser will be coming from a totally different IP address. If I were coding onedrive, I think I'd disallow it as it might look like an attack. Any comments on this? When I'm back in a location where only one IP address is involved, then I'll try this again and report back.

Yes, the issue with not being able to synchronize onedrive appears to be the different VPN domains of the CLI ondrive invocation and the GUI web browser that I was using to copy & paste the URLs. I say this because when the CLI and GUI interfaces were on the same VPN, the same computer, then synchronizing worked fine.

My next step is to attempt using a CLI browser to copy & paste the URLs. 'links' and 'lynx' seem to be the most popular options.

**Comment by abraunegg** [comment]
@gitkevin 
The auth urls should not be different at all - it should be 100% the same regardless how you are doing it.

Your first action however should be to fix your client version.
